### PR TITLE
doc: Add custom runtime with PVC example

### PR DESCRIPTION
**Pull request by zhlsunshine**
#### Motivation

I'm very happy to quickly install `kserve modelmesh` by following [quickstart](https://github.com/kserve/modelmesh-serving/blob/main/docs/quickstart.md), but I encountered the problem when I want to write a python-based custom `ServingRuntime` for model `mnist-svm.joblib` which is also used in many guides and docs, and I also opened an [issue ](https://github.com/kserve/modelmesh-serving/issues/494) for it. However, I can hardly find the complete process which is available online, including the community. So I pieced together some information to make it's easier for user to do it.

#### Modifications

1. Add a REAMD for describing the complete process
2. Add a folder for custom ServingRuntime, including the python code, Dockerfile and required library configuration file.

#### Result

The process for writing a python-based custom `ServingRuntime` is completed, hope it can be a easy start for fresh user.

### CODE REVIEW

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 1]
```diff
@@ -0,0 +1,403 @@
+# The Python-Based Custom Runtime with Model Stored on Persistent Volume Claim
+
+This document can provide the step by step instructions to demonstrate how to write a custom python-based `ServingRuntime` of `MLModel` and store the trained models on persistent volume claims.
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:3]:
```suggestion
This document provides step-by-step instructions to demonstrate how to write a custom Python-based `ServingRuntime` inheriting from [MLServer's MLModel class](https://github.com/SeldonIO/MLServer/blob/master/mlserver/model.py) and deploy a model stored on persistent volume claims with it. 
```

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 2]
```diff
@@ -0,0 +1,403 @@
...
+
+```shell
+kubectl get servingruntimes
+
+NAME             DISABLED   MODELTYPE     CONTAINERS   AGE
+mlserver-1.x                sklearn       mlserver     2m
+ovms-1.x                    openvino_ir   ovms         2m
+torchserve-0.x              pytorch-mar   torchserve   2m
+triton-2.x                  keras         triton       2m
+```
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:28]:
```suggestion
This example assumes that ModelMesh Serving was deployed using the [quickstart guide](https://github.com/kserve/modelmesh-serving/blob/main/docs/quickstart.md). 
```

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 3]
```diff
@@ -0,0 +1,403 @@
...
+
+## 1. Create PV and PVC for storing model file
+
+```shell
+kubectl apply -f - <<EOF
+---
+apiVersion: v1
+kind: PersistentVolume
+metadata:
+  name: my-models-pv
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:46]:
```suggestion
  name: my-models-pvc
```

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 4]
```diff
@@ -0,0 +1,403 @@
...
+
+Verify the configuration setting:
+
+```shell
+kubectl get cm "model-serving-config" -o jsonpath="{.data['config\.yaml']}"
+```
+
+# Implement the Python-based Custom Runtime on MLServer
+
+All necessary resources are in [custom-model](./custom-model), including source code file and Dockerfile.
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:144]:
```suggestion
All of the necessary resources are contained in [custom-model](./custom-model), including the model code, and the Dockerfile.
```

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 5]
```diff
@@ -0,0 +1,403 @@
...
+
+## 1. Implement the API of MLModel
+
+Please understand the `Custom MLModel Template` before writing the real code by following [Python-based Custom Runtime on MLServer](../mlserver_custom.md). Both `load` and `predict` must be implemented to support the custom `ServingRuntime`, and below snippet can be the simplified implementation of `CustomMLModel` for model `mnist-svm.joblib`.
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:148]:
```suggestion
Both `load` and `predict` must be implemented to support this custom `ServingRuntime`. The snippet below provides a simplified implementation of `CustomMLModel` for model `mnist-svm.joblib`. You can read more about it [here](https://github.com/kserve/modelmesh-serving/blob/main/docs/runtimes/mlserver_custom.md).
```

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 6]
```diff
@@ -0,0 +1,403 @@
...
+            outputs = [
+                ResponseOutput(
+                    name = str(input_name[0]),
+                    shape = predictions.shape,
+                    datatype = "INT64",
+                    data=predictions.tolist(),
+                )
+            ],
+        )
+```
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:212]: I think you can just link to the `custom-model.py` here.

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 7]
```diff
@@ -0,0 +1,403 @@
...
+
+## 2. Build custom ServingRuntim image
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:214]:
```suggestion
## 2. Build the custom `ServingRuntime` image
```

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 8]
```diff
@@ -0,0 +1,403 @@
...
+
+User can use `mlserver` or `docker` to help to build the custom `ServingRuntim` image, and the latter is chosen in this doc. The available Dockerfile is below:
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:216]:
```suggestion
You can use [`mlserver`](https://mlserver.readthedocs.io/en/stable/examples/custom/README.html#building-a-custom-image) or `docker` to help to build the custom `ServingRuntime` image, and the latter is done below. 
```

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 9]
```diff
@@ -0,0 +1,403 @@
...
+    MLSERVER__CUSTOM_GRPC_SERVER_SETTINGS='{"grpc.max_metadata_size": "32768"}' \
+    MLSERVER_MODEL_NAME=dummy-model
+
+# With this setting, the implementation field is not required in the model
+# settings which eases integration by allowing the built-in adapter to generate
+# a basic model settings file
+ENV MLSERVER_MODEL_IMPLEMENTATION=custom_model.CustomMLModel
+
+CMD mlserver start $MLSERVER_MODELS_DIR
+```
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:254]: Similar to the model above, you can just link to the `Dockerfile`

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 10]
```diff
@@ -0,0 +1,403 @@
...
+
+Get into the folder [custom-model](./custom-model), then execute command:
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:256]:
```suggestion
To build the image, execute the following command from within the [custom-model](./custom-model) directory.
```

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 11]
```diff
@@ -0,0 +1,403 @@
...
+```
+
+> **Note**: Please use the `--build-arg` to add the http proxy if there is proxy in user's environment, such as: 
+```shell
+docker build --build-arg HTTP_PROXY=http://<DOMAIN-OR-IP>:PORT --build-arg HTTPS_PROXY=http://<DOMAIN-OR-IP>:PORT -t <DOCKER-HUB-ORG>/custom-model-server:0.1 .
+```
+
+## 3. Define and Apply Custom ServingRuntime
+
+Please understand the `Custom ServingRuntime Template` before defining it by following [Python-based Custom Runtime on MLServer](../mlserver_custom.md). Executing below command to create a custom ServingRuntime.
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:270]:
```suggestion
Below, you will create a ServingRuntime using the image built above. You can learn more about the custom `ServingRuntime` template [here](https://github.com/kserve/modelmesh-serving/blob/main/docs/runtimes/mlserver_custom.md#custom-servingruntime-template).
```

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 12]
```diff
@@ -0,0 +1,403 @@
...
+
+NAME                  DISABLED   MODELTYPE      CONTAINERS   AGE
+mlserver-1.x                     sklearn        mlserver     10m
+my-custom-model-0.x              custom_model   mlserver     10m
+ovms-1.x                         openvino_ir    ovms         10m
+torchserve-0.x                   pytorch-mar    torchserve   10m
+triton-2.x                       keras          triton       10m
+```
+
+## 4. Deploy the Inference Service according to custom ServingRuntime
```
Comment (rafvasq) [inline docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:333]:
```suggestion
## 4. Deploy the InferenceService using the custom ServingRuntime
```

**Review by rafvasq — CHANGES_REQUESTED** [review]
Thank you @zhlsunshine! I think having a complete end-to-end example for this will be very helpful. I've done a first-pass and reviewed some edits that I think will benefit the doc. 

I also think that since this is an E2E example along with ready-to-use code, it can be added to `/docs/examples` instead, maybe in its own directory called `/docs/examples/python-custom-runtime`.

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 1 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Done!

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 2 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Done!

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 3 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Hi @rafvasq, this yaml file firstly create persistent volume, then persistent volume claim, so I use the different name for it.

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 4 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Done.

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 5 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Done

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 6 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Yeah, make sense, I link to `custom-model.py`

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 7 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Done

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 9 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Yeah, make sense, link to `custom_mode/Dockerfile`

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 8 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Done

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 10 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Done

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 11 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Done

File: docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 12 above)
Comment (zhlsunshine) [reply docs/runtimes/python-based-custom-runtime-example/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: Done

**Comment by zhlsunshine** [comment]
Hi @rafvasq, thanks for reviewing and comments, I moved this e2e example into `/docs/examples/python-custom-runtime` as you suggested, and change the doc according to your comments.

File: docs/examples/python-custom-runtime/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+
+# Implement the Python-based Custom Runtime on MLServer
+
+All of the necessary resources are contained in [custom-model](./custom-model), including the model code, and the Dockerfile.
+
+## 1. Implement the API of MLModel
+
+Both `load` and `predict` must be implemented to support this custom `ServingRuntime`. The code file [custom_model.py](./custom-model/custom_model.py) provides a simplified implementation of `CustomMLModel` for model `mnist-svm.joblib`. You can read more about it [here](https://github.com/kserve/modelmesh-serving/blob/main/docs/runtimes/mlserver_custom.md).
+
+## 2. Build the custom `ServingRuntime` image
```
Comment (rafvasq) [inline docs/examples/python-custom-runtime/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md:127]: nit for consistency (I think I may have introduced this, sorry) 

```suggestion
## 2. Build the custom ServingRuntime image
```

**Review by rafvasq — CHANGES_REQUESTED** [review]
Thanks @zhlsunshine, this looks great. Just the small nit, run [`make run fmt`](https://github.com/kserve/modelmesh-serving/blob/main/docs/developer.md#formatting-and-linting-code) to get past the Lint check, and rebase.

File: docs/examples/python-custom-runtime/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md (hunk 13 above)
Comment (zhlsunshine) [reply docs/examples/python-custom-runtime/Python-Based-Custom-Runtime-with-Model-Stored-on-PVC.md]: no problem, I added this, removed.

**Comment by zhlsunshine** [comment]
> Thanks @zhlsunshine, this looks great. Just the small nit, run [`make run fmt`](https://github.com/kserve/modelmesh-serving/blob/main/docs/developer.md#formatting-and-linting-code) to get past the Lint check, and rebase.

Hi @rafvasq  I have rebased code for the PR branch, however, it seems that there may be some problem with the `make run fmt` command itself which is checking the lint in docker container.

**Comment by zhlsunshine** [comment]
Hi @rafvasq, I also completed the `golangci-lint` process, submitted the formatted docs.

**Review by rafvasq — APPROVED** [review]
Thanks @zhlsunshine!

**Pull request merged**